targets/unity-v2/Testing/Tests/Shared/Uunit/UUnitTestContext.cs

```diff
@@ -44,6 +44,29 @@ public UUnitTestContext(UUnitTestCase testInstance, Action<UUnitTestContext> tes
 
         public void EndTest(UUnitFinishState finishState, string resultMsg)
         {
+            if (FinishState != UUnitFinishState.PENDING)
+            {
+                resultMsg += "You can't mark a test as finished twice.\n";
+                if (FinishState == UUnitFinishState.PASSED)
+                {
+                    if (finishState != UUnitFinishState.FAILED)
+                    {
+                        TestResultMsg = resultMsg;
+                        return;
+                    }
+                    resultMsg = TestResultMsg + resultMsg;
+                }
+                else
+                {
+                    if (finishState != UUnitFinishState.FAILED)
+                    {
+                        resultMsg += "You can't mark a test as finished twice.\n";
+                    }
+                    TestResultMsg = resultMsg;
+                    return;
+                }
+            }
```

Comment:
It's really hard to follow this logic.

I would prefer:

if (condition && condition)
// logic
else if (condition && condition)
// logic
else if (condition && condition)
// logic
else if (condition
// logic
else
// logic

Basically, your logic above may be right, but if exactly one state is wrong, then we have to reinvent this whole code. The switch statement I describe above may be more code, but it's a lot easier to debug and add a condition pair that may not be covered by your logic.